api/paidAction/index.js

```diff
@@ -107,283 +202,286 @@ export default async function performPaidAction (actionType, args, context) {
-      invoice,
-      paymentMethod: 'OPTIMISTIC'
+    const subContext = { ...context, tx }
+    const failedInvoice = await models.invoice.findUnique({ where: { id: invoiceId, userId: me?.id ?? USER_ID.anon } })
+    if (!failedInvoice) {
+      throw new Error('invoice not found')
     }
-  }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
-}
-
-const INVOICE_EXPIRE_SECS = 600
-const MAX_PENDING_PAID_ACTIONS_PER_USER = 100
 
-export async function assertBelowMaxPendingInvoices (context) {
-  const { models, me } = context
-  const pendingInvoices = await models.invoice.count({
-    where: {
-      userId: me?.id ?? USER_ID.anon,
-      actionState: {
-        notIn: PAID_ACTION_TERMINAL_STATES
-      }
+    if (
+      failedInvoice.actionState !== 'FAILED' &&
+      failedInvoice.actionState !== 'PENDING' &&
+      failedInvoice.actionState !== 'PENDING_HELD'
+    ) {
+      throw new Error(`actions is not in a retriable state: ${failedInvoice.actionState}`)
     }
-  })
 
-  if (pendingInvoices >= MAX_PENDING_PAID_ACTIONS_PER_USER) {
-    throw new Error('You have too many pending paid actions, cancel some or wait for them to expire')
-  }
-}
+    // cancel the previous invoice
+    await finalizeHodlInvoice({ data: { hash: failedInvoice.hash }, lnd, models, boss })
+    const actionType = failedInvoice.actionType
 
-export class NonInvoiceablePeerError extends Error {
-  constructor () {
-    super('non invoiceable peer')
-    this.name = 'NonInvoiceablePeerError'
-  }
-}
+    const paidAction = paidActions[actionType]
+    if (!paidAction) {
+      throw new Error(`invalid action type ${actionType}`)
+    }
 
-export async function createWrappedInvoice (actionType, args, context) {
-  // if the action has an invoiceable peer, we'll create a peer invoice
-  // wrap it, and return the wrapped invoice
-  const { cost, models, lnd, me } = context
-  const userId = await paidActions[actionType]?.invoiceablePeer?.(args, context)
-  if (!userId) {
-    throw new NonInvoiceablePeerError()
-  }
+    if (!failedInvoice) {
+      throw new Error(`missing invoice ${actionType}`)
+    }
 
-  await assertBelowMaxPendingInvoices(context)
+    const { msatsRequested, actionId, actionArgs } = failedInvoice
+    subContext.cost = msatsRequested
+    subContext.actionId = actionId
+    subContext.retryForInvoice = failedInvoice
+    subContext.forceInternal = forceInternal
+    subContext.actionAttempt = failedInvoice.actionAttempt + 1 // next attempt
+    subContext.prioritizeInternal = prioritizeInternal
 
-  const description = await paidActions[actionType].describe(args, context)
-  const { invoice: bolt11, wallet } = await createUserInvoice(userId, {
-    // this is the amount the stacker will receive, the other 3/10ths is the sybil fee
-    msats: cost * BigInt(7) / BigInt(10),
-    description,
-    expiry: INVOICE_EXPIRE_SECS
-  }, { models })
+    if (resetAttempts) {
```

Comment:
this is passes when the user clicks the `retry` button: we reset the cycle and we start over